jwks.go

```diff
@@ -156,6 +156,89 @@ func (r *remoteKeySet) keysWithID(ctx context.Context, keyIDs []string) ([]jose.
 	return keys, nil
 }
 
+func (r *remoteKeySet) keysWithAlgsFromCache(algs []string) ([]jose.JSONWebKey, bool) {
+	r.mu.Lock()
+	keys, expiry := r.cachedKeys, r.expiry
+	r.mu.Unlock()
+
+	// Have the keys expired?
+	if expiry.Add(keysExpiryDelta).Before(r.now()) {
+		return nil, false
+	}
+
+	var signingKeys []jose.JSONWebKey
+	for _, key := range keys {
+		if contains(algs, key.Algorithm) {
+			signingKeys = append(signingKeys, key)
+		}
+	}
+
+	if len(signingKeys) == 0 {
+		// Are the keys about to expire?
+		if r.now().Add(keysExpiryDelta).After(expiry) {
+			return nil, false
+		}
+	}
+
+	return signingKeys, true
+}
+func (r *remoteKeySet) keysWithAlgs(ctx context.Context, algs []string) ([]jose.JSONWebKey, error) {
+	keys, ok := r.keysWithAlgsFromCache(algs)
+	if ok {
+		return keys, nil
+	}
+
+	var inflightCtx *inflight
+	func() {
+		r.mu.Lock()
```

Comment:
I really dislike replicating this logic. It's fragile because it's dealing with the retry/sync mechanism.